contracts/src/registry/L1ETHRegistry.sol

```diff
@@ -0,0 +1,244 @@
+     */
+    function ejectFromL2(string calldata label, address owner, IRegistry registry, uint96 flags, uint64 expires)
+        public
+        onlyRole(EJECTION_CONTROLLER_ROLE)
+        returns (uint256 tokenId)
+    {
+        tokenId = (uint256(keccak256(bytes(label))) & ~uint256(FLAGS_MASK)) | flags;
+        flags = (flags & FLAGS_MASK) | (uint96(expires) << 32);
+
+        // If there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        datastore.setSubregistry(tokenId, address(registry), flags);
+        emit NewSubname(label);
+        emit NameEjected(tokenId, owner, expires);
+        return tokenId;
+    }
+
+    /**
+     * @dev Renew an ejected name
+     * @param tokenId The token ID of the name to renew
+     * @param expires New expiration timestamp
+     */
+    function renew(uint256 tokenId, uint64 expires) public onlyRole(RENEWAL_CONTROLLER_ROLE) {
+        (address subregistry, uint96 flags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiration = _extractExpiry(flags);
+        if (oldExpiration < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        if (expires < oldExpiration) {
+            revert CannotReduceExpiration(oldExpiration, expires);
+        }
+        datastore.setSubregistry(tokenId, subregistry, (flags & FLAGS_MASK) | (uint96(expires) << 32));
+
+        address observer = tokenObservers[tokenId];
+        if (observer != address(0)) {
+            L1ETHRegistryTokenObserver(observer).onRenew(tokenId, expires, msg.sender);
+        }
+
+        emit NameRenewed(tokenId, expires, msg.sender);
+    }
+
+    /**
+     * @dev Set the token observer for a name
+     * @param tokenId The token ID of the name
+     * @param _observer The observer address
+     */
+    function setTokenObserver(uint256 tokenId, address _observer) external onlyTokenOwner(tokenId) {
+        tokenObservers[tokenId] = _observer;
+        emit TokenObserverSet(tokenId, _observer);
+    }
+
+    /**
+     * @dev Relinquish an ejected name
+     * @param tokenId The token ID of the name to relinquish
+     */
+    function relinquish(uint256 tokenId) external onlyTokenOwner(tokenId) {
+        _burn(ownerOf(tokenId), tokenId, 1);
+        datastore.setSubregistry(tokenId, address(0), 0);
+
+        address observer = tokenObservers[tokenId];
+        if (observer != address(0)) {
+            L1ETHRegistryTokenObserver(observer).onRelinquish(tokenId, msg.sender);
+        }
+
+        emit NameRelinquished(tokenId, msg.sender);
+    }
+
+    /**
+     * @dev Migrate a name back to L2
+     * @param tokenId The token ID of the name to migrate
+     * @param sendTo The address to send the name to on L2
+     */
+    function migrateToL2(uint256 tokenId, address sendTo)
```

Comment:
This needs to notify the ejection controller, so that control can be given back to the user on L2.

Reply:
Added TODO for now.